packages/vitest/src/node/pools/browser.ts

```diff
@@ -38,7 +38,7 @@ export function createBrowserPool(ctx: Vitest): ProcessPool {
     const provider = project.browserProvider!
     providers.add(provider)
 
-    const origin = `http://${ctx.config.browser.api?.host || 'localhost'}:${project.browser!.config.server.port}`
```

Comment:
There is a possibility that this might be null if server is restarted (due to a config change): https://github.com/vitejs/vite/pull/15450

Does it affect us?

Reply:
Thanks for the reference. There could be something sketchy scenario. I'll look into it.

Also another thing, actually it might not be preferred to use `resolveUrls` since it looks like Vite has some inconsistency when users set vite config like `server.host: "local.example.com"`.

- https://github.com/vitejs/vite/issues/11342
- https://github.com/vitejs/vite/issues/8200

Let me think about it.

Reply:
I just remembered this issue about Vite's `--open` default behavior:
- https://github.com/vitejs/vite/issues/14921#issuecomment-1805655613

Essentially Vite does a following when `ViteDevServer.openBrowser`:

```tsx
new URL(options.open, resolvedUrls.local[0] ?? resolvedUrls.network[0])
```

https://github.com/vitejs/vite/blob/2687dbbd4e19c86f9888ee784c9b51598e8b79ca/packages/vite/src/node/server/index.ts#L484-L492

This also means that `provider: "none"` is already forcing url origin to be `resolvedUrls.local[0] ?? resolvedUrls.network[0]`, so aligning this behavior for all providers might not be wrong.

Reply:
Regarding Vite server's restart behavior, I don't think that's relevant for Vitest browser mode since browser mode uses a dedicated Vite dev server, which doesn't watch files on its own and is completely re-spawned when Vitest's main server restarts:

https://github.com/vitest-dev/vitest/blob/b2c79e4431ee0bdaa18484aad365fc0fca267b3f/packages/vitest/src/integrations/browser/server.ts#L21-L27

https://github.com/vitest-dev/vitest/blob/b2c79e4431ee0bdaa18484aad365fc0fca267b3f/packages/vitest/src/node/workspace.ts#L271-L276

So, relying on `resolvedUrls` should be safe before landing https://github.com/vitejs/vite/pull/15450 in Vite v5.1.

Currently the log looks like this for `playwright` provider:

```
Restarting due to config changes...

 DEV  v1.2.0 /home/hiroshi/code/others/vitest/test/browser/fixtures/server-url
      Browser runner started at http://localhost:5174/
```

But `webdriverio` actually doesn't support restart at all since it kills a whole process when Vitest main server restarts and closes all pools to recreate:

https://github.com/vitest-dev/vitest/blob/b2c79e4431ee0bdaa18484aad365fc0fca267b3f/packages/vitest/src/node/core.ts#L88

https://github.com/vitest-dev/vitest/blob/b2c79e4431ee0bdaa18484aad365fc0fca267b3f/packages/browser/src/node/providers/webdriver.ts#L93-L95